File: 2021/day6.py
import sys


def main():
    lines = [line.rstrip() for line in sys.stdin]
    data = {k: 0 for k in range(1, 6)}
    for n in map(int, lines[0].split(",")):
        data[n] += 1

    # Part 1
    for _ in range(80):
        data = step(data)
    print(sum(data.values()))

    # Part 2
    for _ in range(256 - 80):
        data = step(data)
    print(sum(data.values()))


def step(data):
    next_data = {k: 0 for k in range(1, 9)}
    for k, v in data.items():
        if v == 0:
            continue
        if k == 0:
            next_data[6] += v
            next_data[8] += v
        else:
            next_data[k - 1] = v
    return next_data


if __name__ == "__main__":
    main()

File: 2021/rust/src/bin/day6.rs
use std::io;

fn main() {
    let mut text = String::new();
    io::stdin().read_line(&mut text).unwrap();
    let mut data = vec![0u64; 9];
    for c in text.strip_suffix("\n").unwrap().split(",") {
        data[str::parse::<usize>(c).unwrap()] += 1;
    }
    for _ in 0..80 {
        step(&mut data);
    }
    println!("{}", data.iter().sum::<u64>());
    for _ in 80..256 {
        step(&mut data);
    }
    println!("{}", data.iter().sum::<u64>());
}

fn step(data: &mut Vec<u64>) {
    let d0 = data[0];
    for d in 0..=7 {
        data[d] = data[d + 1];
    }
    data[6] += d0;
    data[8] = d0;
}
